### PR TITLE
docs: Update XDP drivers list

### LINKED ISSUE #22867: Update XDP driver list in docs
## Cilium Feature Proposal

I didn't see a way to submit a docs request so I am submitting this as a CFP. I hope that's okay.

**Is your feature request related to a problem?**

The drivers listed in https://github.com/cilium/cilium/blob/v1.12/Documentation/bpf.rst#xdp section `Drivers supporting native XDP` is from linux 4.17. 

**Describe the feature you'd like**

It's unclear how this list was generated and it would be nice if it was updated for a recent kernel.

### PR DESCRIPTION
**Pull request by janvi01**
Please ensure your pull request adheres to the following guidelines:

- [X] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

Fixes: #22867 

```release-note
Update XDP drivers support list in BPF docs
```

moved the Driver supporting XDP list to a dedicated RST fragment to include it from where it has to be listed.

### CODE REVIEW

**Comment by janvi01** [comment]
@qmonnet

**Review by ghost — COMMENTED** [review]
I'm not sure what's the benefit from this change? The drivers table appears in docs only once and it's not autogenerated, so I don't see why it's better as a separate fragment. Also, why change the table format to follow the bcc docs? To me the existing table format seems easier to comprehend.

**Review by qmonnet — CHANGES_REQUESTED** [review]
I agree with Anna, it's no longer necessary to move the table since we only have one list in the docs today. Sorry for misleading you about this earlier.

And I don't know why we would copy the list from the BCC docs, worst of all with no attribution.

What we need to do to address #22867 is to keep the current list and format, but update the list with the drivers who recently got support for XDP in the Linux kernel driver. This `grep` command you found should help you determine what drivers have support.

**Comment by janvi01** [comment]
> What we need to do to address #22867 is to keep the current list and format, but update the list with the drivers who recently got support for XDP in the Linux kernel driver. This `grep` command you found should help you determine what drivers have support.

Thanks @qmonnet for clarifying, I will do the required changes.

File: Documentation/bpf/progtypes.rst [hunk 1]
```diff
@@ -338,49 +338,57 @@ corresponding network driver name of an interface can be determined as follows:
...
++-------------------+---------------+-------------+
+| Broadcom          | bnxt_en       | >= 4.11     |
++-------------------+---------------+-------------+
+| Cavium            | qed           | >= 4.10     |
+|                   +---------------+-------------+
+|                   | thunderx      | >= 4.12     |
++-------------------+---------------+-------------+
+| Freescale         | dpaa2         | >= 5.0      |
++-------------------+---------------+-------------+
+| Generic XDP       |               | >= 4.12     |
```
Comment (gentoo-root) [inline Documentation/bpf/progtypes.rst]: The text above says:

> A list of drivers supporting native XDP can be found in the table below.

I think it's confusing to include generic XDP in the table of native XDP drivers (it's not native, and it's not a driver).

**Review by gentoo-root — CHANGES_REQUESTED** [review]

File: Documentation/bpf/progtypes.rst (hunk 1 above)
Comment (janvi01) [reply Documentation/bpf/progtypes.rst]: ah yeah, better to not list it here. Thanks, updated.

**Review by gentoo-root — APPROVED** [review]
Looks good from my side. Please squash all commits before merging and update the release note in the pull request description.

File: Documentation/bpf/progtypes.rst [hunk 2]
```diff
@@ -338,49 +338,55 @@ corresponding network driver name of an interface can be determined as follows:
...
-| Texas Instruments | cpsw       | >= 5.3      |
-+-------------------+------------+-------------+
++-------------------+---------------+-------------+
+| Vendor            | Driver        | XDP Support |
++===================+===============+=============+
+| Amazon            | ena           | >= 5.6      |
++-------------------+---------------+-------------+
+| Broadcom          | bnxt_en       | >= 4.11     |
++-------------------+---------------+-------------+
+| Cavium            | qed           | >= 4.10     |
```
Comment (qmonnet) [inline Documentation/bpf/progtypes.rst]: Where does this one come from?

File: Documentation/bpf/progtypes.rst [hunk 3]
```diff
@@ -338,49 +338,55 @@ corresponding network driver name of an interface can be determined as follows:
...
+|                   +---------------+-------------+
+|                   | thunderx      | >= 4.12     |
++-------------------+---------------+-------------+
+| Freescale         | dpaa2         | >= 5.0      |
++-------------------+---------------+-------------+
+| Intel             | ice           | >= 5.5      |
+|                   +---------------+-------------+
+|                   | igb           | >= 5.10     |
```
Comment (qmonnet) [inline Documentation/bpf/progtypes.rst]: Looks good

**Review by qmonnet — CHANGES_REQUESTED** [review]
There seems to be some drivers still missing, can you please complete the list?

**Comment by janvi01** [comment]
> There seems to be some drivers still missing, can you please complete the list?

Added the new ones that were missing, which one is left?

File: Documentation/bpf/progtypes.rst (hunk 2 above)
Comment (janvi01) [reply Documentation/bpf/progtypes.rst]: QLogic (Cavium) qed* drivers | 4.10 | 496e05170958
-- | -- | --

Comment (gentoo-root) [reply Documentation/bpf/progtypes.rst]: > Where does this one come from?

QLogic is Cavium's acquisition. I'm not familiar with the structure of their driver (i.e. which subdirectories are based on which), but XDP parts are there in **qede** in 4.10:

https://elixir.bootlin.com/linux/v4.10/source/drivers/net/ethernet/qlogic/qede/qede_main.c#L2528

In the recent kernels it looks like XDP for PFs is supported by all subdrivers of **qede**, and XDP for VFs is only supported by **qed**:

https://elixir.bootlin.com/linux/v6.7/source/drivers/net/ethernet/qlogic/qed/qed_l2.c#L2254
https://elixir.bootlin.com/linux/v6.7/source/drivers/net/ethernet/qlogic/qed/qed_l2.c#L2273
https://elixir.bootlin.com/linux/v6.7/source/drivers/net/ethernet/qlogic/qede/qede_main.c#L856

Comment (gentoo-root) [reply Documentation/bpf/progtypes.rst]: > QLogic is Cavium's acquisition.

I wonder if we should rename Mellanox to NVIDIA, or stick with QLogic, or provide both names in both cases.

**Comment by qmonnet** [comment]
```
$ rg -l XDP_SETUP_PROG drivers/net/ethernet | sort
drivers/net/ethernet/amazon/ena/ena_xdp.c
drivers/net/ethernet/aquantia/atlantic/aq_main.c
drivers/net/ethernet/broadcom/bnxt/bnxt_xdp.c
drivers/net/ethernet/cavium/thunder/nicvf_main.c
drivers/net/ethernet/engleder/tsnep_main.c
drivers/net/ethernet/freescale/dpaa2/dpaa2-eth.c
drivers/net/ethernet/freescale/dpaa/dpaa_eth.c
drivers/net/ethernet/freescale/enetc/enetc.c
drivers/net/ethernet/freescale/fec_main.c
drivers/net/ethernet/fungible/funeth/funeth_main.c
drivers/net/ethernet/google/gve/gve_main.c
drivers/net/ethernet/intel/i40e/i40e_main.c
drivers/net/ethernet/intel/ice/ice_main.c
drivers/net/ethernet/intel/igb/igb_main.c
drivers/net/ethernet/intel/igc/igc_main.c
drivers/net/ethernet/intel/ixgbe/ixgbe_main.c
drivers/net/ethernet/intel/ixgbevf/ixgbevf_main.c
drivers/net/ethernet/marvell/mvneta.c
drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
drivers/net/ethernet/mediatek/mtk_eth_soc.c
drivers/net/ethernet/mellanox/mlx4/en_netdev.c
drivers/net/ethernet/mellanox/mlx5/core/en_main.c
drivers/net/ethernet/microchip/lan966x/lan966x_xdp.c
drivers/net/ethernet/microsoft/mana/mana_bpf.c
drivers/net/ethernet/netronome/nfp/nfp_net_common.c
drivers/net/ethernet/qlogic/qede/qede_filter.c
drivers/net/ethernet/sfc/efx.c
drivers/net/ethernet/sfc/siena/efx.c
drivers/net/ethernet/socionext/netsec.c
drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
drivers/net/ethernet/ti/cpsw_priv.c
```

We don't have all of these in the list. I'm not sure every file listed above corresponds to a distinct driver, but there are definitely a few that implement XDP and that are not in our list.

File: Documentation/bpf/progtypes.rst (hunk 2 above)
Comment (qmonnet) [reply Documentation/bpf/progtypes.rst]: > QLogic is Cavium's acquisition.

Ah right now I see, thanks Maxim.

But we already have an entry for QLogic, so adding Cavium creates a duplicate.

> I wonder if we should rename Mellanox to NVIDIA, or stick with QLogic, or provide both names in both cases.

Good question. I'd stick with the current names, that are also easier to map to kernel drivers, and change if the names change in the kernel tree?

Comment (gentoo-root) [reply Documentation/bpf/progtypes.rst]: > But we already have an entry for QLogic, so adding Cavium creates a duplicate.

Oh right, there is one down below. We don't need to add Cavium/qed then.

> I'd stick with the current names, that are also easier to map to kernel drivers, and change if the names change in the kernel tree?

Makes sense to me.

**Comment by janvi01** [comment]
@qmonnet @gentoo-root Just updated the list, I have a few ones that I am confused to add or not.

```
Likely Support XDP (Based on Documentation or Other Indications):

drivers/net/ethernet/marvell/mvneta.c (mvneta driver by Marvell)
drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c (mvpp2 driver by Marvell)
drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c (OcteonTx2 PF driver by Marvell)
drivers/net/ethernet/mediatek/mtk_eth_soc.c (MTK Ethernet SoC driver by MediaTek)
```

**Comment by janvi01** [comment]
@gentoo-root The list is completed.

File: Documentation/bpf/progtypes.rst [hunk 4]
```diff
@@ -338,49 +338,92 @@ corresponding network driver name of an interface can be determined as follows:
...
+|                        +-------------------------+-------------+
+|                        | tun/tap                 | >= 4.14     |
+|                        +-------------------------+-------------+
+|                        | virtio_net              | >= 4.10     |
+|                        +-------------------------+-------------+
+|                        | xen-netfront            | >= 5.9      |
+|                        +-------------------------+-------------+
+|                        | Veth                    | >= 4.19     |
++------------------------+-------------------------+-------------+
+| Qlogic                 | qede                    | >= 4.10     |
```
Comment (qmonnet) [inline Documentation/bpf/progtypes.rst]:
```suggestion
| QLogic                 | qede                    | >= 4.10     |
```

File: Documentation/bpf/progtypes.rst [hunk 5]
```diff
@@ -338,49 +338,92 @@ corresponding network driver name of an interface can be determined as follows:
...
++------------------------+-------------------------+-------------+
+| SFC                    | Efx                     | >= 5.5      |
```
Comment (qmonnet) [inline Documentation/bpf/progtypes.rst]: Duplicate with Solarflare's sfc?

File: Documentation/bpf/progtypes.rst [hunk 6]
```diff
@@ -338,49 +338,92 @@ corresponding network driver name of an interface can be determined as follows:
...
++------------------------+-------------------------+-------------+
+| Mediatek               | mtk                     | >= 6.0      |
++------------------------+-------------------------+-------------+
+| Mellanox               | mlx4                    | >= 4.8      |
+|                        +-------------------------+-------------+
+|                        | mlx5                    | >= 4.9      |
++------------------------+-------------------------+-------------+
+| Microchip              | lan966x                 | >= 6.1      |
++------------------------+-------------------------+-------------+
+| Microsoft              | hv_netvsc               | >= 5.6      |
```
Comment (qmonnet) [inline Documentation/bpf/progtypes.rst]:
```suggestion
| Microsoft              | hv_netvsc (Hyper-V)     | >= 5.6      |
```

File: Documentation/bpf/progtypes.rst [hunk 7]
```diff
@@ -338,49 +338,92 @@ corresponding network driver name of an interface can be determined as follows:
...
++------------------------+-------------------------+-------------+
+| Solarflare             | sfc                     | >= 5.5      |
++------------------------+-------------------------+-------------+
+| STMicro                | stmmac                  | >= 5.3      |
++------------------------+-------------------------+-------------+
+| Texas Instruments      | cpsw                    | >= 5.3      |
++------------------------+-------------------------+-------------+
+| VMware                 | vmxnet3                 | >= 6.6      |
++------------------------+-------------------------+-------------+
+| Xen                    | xennet                  | >= 5.9      |
```
Comment (qmonnet) [inline Documentation/bpf/progtypes.rst]: Isn't that one the same as `xen-netfront` in the “Others” section?

**Review by qmonnet — CHANGES_REQUESTED** [review]

File: Documentation/bpf/progtypes.rst (hunk 5 above)
Comment (janvi01) [reply Documentation/bpf/progtypes.rst]: yes, just checked.

File: Documentation/bpf/progtypes.rst (hunk 7 above)
Comment (janvi01) [reply Documentation/bpf/progtypes.rst]: ah yes, both are same.

**Comment by janvi01** [comment]
Thanks, @qmonnet I made the required changes.

**Review by qmonnet — COMMENTED** [review]
OK, looks like we're getting here, thanks!

Can you please squash all your commits into a single one, and force-push onto the branch?

File: Documentation/bpf/progtypes.rst [hunk 8]
```diff
@@ -338,49 +338,88 @@ corresponding network driver name of an interface can be determined as follows:
...
-| Solarflare        | sfc        | >= 5.5      |
-+-------------------+------------+-------------+
-| Texas Instruments | cpsw       | >= 5.3      |
-+-------------------+------------+-------------+
++------------------------+-------------------------+-------------+
+| Vendor                 | Driver                  | XDP Support |
++========================+=========================+=============+
+| Amazon                 | ena                     | >= 5.6      |
++------------------------+-------------------------+-------------+
+| Aquantia               | atlantic                | >= 5.13     |
```
Comment (dylandreimerink) [inline Documentation/bpf/progtypes.rst]: If I am not mistaken, support for atlantic was added in v5.19 with this PR https://github.com/torvalds/linux/commit/0d14657f40830243266f972766f1e4d00436e648

File: Documentation/bpf/progtypes.rst [hunk 9]
```diff
@@ -338,49 +338,88 @@ corresponding network driver name of an interface can be determined as follows:
...
+| Broadcom               | bnxt_en                 | >= 4.11     |
++------------------------+-------------------------+-------------+
+| Cavium                 | thunderx                | >= 4.12     |
++------------------------+-------------------------+-------------+
+| Engleder               | tsne                    | >= 6.3      |
+|                        | (TSN Express Path)      |             |
++------------------------+-------------------------+-------------+
+| Freescale              | dpaa                    | >= 5.0      |
+|                        +-------------------------+-------------+
+|                        | dpaa2                   | >= 5.11     |
```
Comment (dylandreimerink) [inline Documentation/bpf/progtypes.rst]: The versions are reversed dpaa2 got support in 5.0 [see](https://github.com/torvalds/linux/commit/7e273a8ebdd3b83f94eb8b49fc8ee61464f47cc2)
and dpaa got support in v5.11 [see](https://github.com/torvalds/linux/commit/86c0c196cbe48f844721783d9162e46bc35c0c5a)

File: Documentation/bpf/progtypes.rst [hunk 10]
```diff
@@ -338,49 +338,88 @@ corresponding network driver name of an interface can be determined as follows:
...
+|                        +-------------------------+-------------+
+|                        | enetc                   | >= 5.13     |
+|                        +-------------------------+-------------+
+|                        | fec_enet                | >= 6.2      |
++------------------------+-------------------------+-------------+
+| Fungible               | fun                     | >= 5.18     |
++------------------------+-------------------------+-------------+
+| Google                 | gve                     | >= 5.19     |
```
Comment (dylandreimerink) [inline Documentation/bpf/progtypes.rst]: If I am not mistaken support for gve was added in v6.4 [see](https://github.com/torvalds/linux/commit/75eaae158b1b7d8d5bde2bafc0bcf778423071d3)

File: Documentation/bpf/progtypes.rst [hunk 11]
```diff
@@ -338,49 +338,88 @@ corresponding network driver name of an interface can be determined as follows:
...
+|                        +-------------------------+-------------+
+|                        | otx2                    | >= 5.16     |
++------------------------+-------------------------+-------------+
+| Mediatek               | mtk                     | >= 6.0      |
++------------------------+-------------------------+-------------+
+| Mellanox               | mlx4                    | >= 4.8      |
+|                        +-------------------------+-------------+
+|                        | mlx5                    | >= 4.9      |
++------------------------+-------------------------+-------------+
+| Microchip              | lan966x                 | >= 6.1      |
```
Comment (dylandreimerink) [inline Documentation/bpf/progtypes.rst]: I believe for lan966x support landed in 6.2 [see](https://github.com/torvalds/linux/commit/6a2159be7604f5cdd7f574f4e0922f61e63c3f16)

File: Documentation/bpf/progtypes.rst [hunk 12]
```diff
@@ -338,49 +338,88 @@ corresponding network driver name of an interface can be determined as follows:
...
++------------------------+-------------------------+-------------+
+| Microsoft              | hv_netvsc (Hyper-V)     | >= 5.6      |
+|                        +-------------------------+-------------+
+|                        | mana                    | >= 6.0      |
```
Comment (dylandreimerink) [inline Documentation/bpf/progtypes.rst]: If I am not mistaken mana was added in v5.17 [see](https://github.com/torvalds/linux/commit/ed5356b53f070dea5dff5a01b740561cb8222199)

File: Documentation/bpf/progtypes.rst [hunk 13]
```diff
@@ -338,49 +338,88 @@ corresponding network driver name of an interface can be determined as follows:
...
+|                        +-------------------------+-------------+
+|                        | Veth                    | >= 4.19     |
++------------------------+-------------------------+-------------+
+| QLogic                 | qede                    | >= 4.10     |
++------------------------+-------------------------+-------------+
+| Socionext              | netsec                  | >= 5.3      |
++------------------------+-------------------------+-------------+
+| Solarflare             | SFC Efx                 | >= 5.5      |
++------------------------+-------------------------+-------------+
+| STMicro                | stmmac                  | >= 5.3      |
```
Comment (dylandreimerink) [inline Documentation/bpf/progtypes.rst]: Support for stmmac added in v5.13 [see](https://github.com/torvalds/linux/commit/5fabb01207a2d3439a6abe1d08640de9c942945f)

**Review by dylandreimerink — CHANGES_REQUESTED** [review]
Sorry for the late review. I believe there are several incorrect version numbers. 

The version numbers in my comments are based on the commits where `ndo_bpf` hooks were added, which I believe to be accurate and the first branch which includes this commit.

Not sure which source was used for these existing numbers, perhaps my method is flawed, in which case I would like to be enlightened.

**Comment by janvi01** [comment]
Thanks @dylandreimerink for the revised and accurate versions. my apologies for not giving accurate.

**Review by dylandreimerink — APPROVED** [review]
Thanks for the version fixes!

**Comment by janvi01** [comment]
oops, I am facing an issue with squashing my commits into a single one. @qmonnet

File: Documentation/spelling_wordlist.txt [hunk 14]
```diff
@@ -93,6 +97,7 @@ Pfaff
 PodCIDR
 Polytechnique
 QCon
+QLogic
 Qlogic
```
Comment (gentoo-root) [inline Documentation/spelling_wordlist.txt]: Remove Qlogic if we renamed it to QLogic?

File: Documentation/bpf/progtypes.rst [hunk 15]
```diff
@@ -338,49 +338,88 @@ corresponding network driver name of an interface can be determined as follows:
...
+|                        +-------------------------+-------------+
+|                        | netdevsim               | >= 4.16     |
+|                        +-------------------------+-------------+
+|                        | tun/tap                 | >= 4.14     |
+|                        +-------------------------+-------------+
+|                        | virtio_net              | >= 4.10     |
+|                        +-------------------------+-------------+
+|                        | xen-netfront            | >= 5.9      |
+|                        +-------------------------+-------------+
+|                        | Veth                    | >= 4.19     |
```
Comment (gentoo-root) [inline Documentation/bpf/progtypes.rst]: Why capitalize veth? I think it's normally all lower case letters.

**Review by dylandreimerink — APPROVED** [review]

**Comment by qmonnet** [comment]
Squashed the last fixes and force-pushed.

**Review by qmonnet — APPROVED** [review]

**Pull request merged**